pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1202 @@ func (o *ClusterUninstaller) validate() error {
+	segments := strings.SplitN(roleARN.Resource, "/", 2)
+	if len(segments) != 2 {
+		return errors.Errorf("cannot parse resource %q into {type}/{id}", roleARN.Resource)
+	}
+	resourceType := segments[0]
+	name := segments[1]
+	logger = logger.WithField("name", name)
+
+	if resourceType != "role" {
+		return errors.Errorf("%s ARN passed to deleteIAMRole: %s", resourceType, roleARN.String())
+	}
 
-	var result *route53.ListTagsForResourceOutput
 	var err error
-	for _, zone := range zones {
-		for {
-			result, err = r53Client.ListTagsForResource(&route53.ListTagsForResourceInput{
-				ResourceType: aws.String("hostedzone"),
-				ResourceId:   zone.Id,
-			})
-			if err != nil {
-				if request.IsErrorThrottle(err) {
-					logger.Debugf("sleeping before trying to resolve tags for zone %s: %v", *zone.Id, err)
-					time.Sleep(time.Second)
-					continue
+	err2 := client.ListRolePoliciesPages(
+		&iam.ListRolePoliciesInput{RoleName: &name},
+		func(results *iam.ListRolePoliciesOutput, lastPage bool) bool {
+			for _, policy := range results.PolicyNames {
+				_, err = client.DeleteRolePolicy(&iam.DeleteRolePolicyInput{
+					RoleName:   &name,
+					PolicyName: policy,
+				})
+				if err != nil {
+					err = errors.Wrapf(err, "deleting IAM role policy %s", *policy)
+					logger.Info(err)
 				}
-				return zonesAsAWSObjects, err
+				logger.WithField("policy", *policy).Info("Deleted")
 			}
 
-			break
-		}
-
-		tagsToMap, err := tagsToMap(result.ResourceTagSet.Tags)
-		if err != nil {
-			return zonesAsAWSObjects, err
-		}
-
-		zonesAsAWSObjects = append(zonesAsAWSObjects, awsObjectWithTags{
-			Name: *zone.Id,
-			Tags: tagsToMap,
-		})
+			return !lastPage
+		},
+	)
 
+	if err != nil {
+		return err
+	}
+	if err2 != nil {
+		return errors.Wrap(err2, "listing IAM role policies")
 	}
 
-	return zonesAsAWSObjects, nil
-}
+	err2 = client.ListInstanceProfilesForRolePages(
+		&iam.ListInstanceProfilesForRoleInput{RoleName: &name},
+		func(results *iam.ListInstanceProfilesForRoleOutput, lastPage bool) bool {
+			for _, profile := range results.InstanceProfiles {
+				parsed, err := arn.Parse(*profile.Arn)
+				if err != nil {
+					err = errors.Wrap(err, "parse ARN for IAM instance profile")
+					logger.Info(err)
+					continue
+				}
 
-// deleteEntriesFromSharedR53Zone will find route53 entries for the shared (ie non-terraform-managed) route53 zone
-// and remove them.
-// Provide the terraform-created private zone, and the manually created public/shared zone, and it will find any
-// entries in the public/shared zone that match entries in the private zone, and delete them
-func deleteEntriesFromSharedR53Zone(zoneID string, sharedZoneID string, r53Client *route53.Route53, logger logrus.FieldLogger) error {
+				err = deleteIAMInstanceProfile(client, parsed, logger.WithField("IAM instance profile", parsed.String()))
+				if err != nil {
+					err = errors.Wrapf(err, "deleting %s", parsed.String())
+					logger.Info(err)
+				}
+			}
+
+			r
```

Comment:
This pull request adds pagination for a number of resources that were missing pagination before (and even when we called `*Pages()` before, [we were doing it wrong][1]):

```console
$ git grep 'Pages(' pkg/destroy/aws/aws.go
pkg/destroy/aws/aws.go:				err = tagClient.GetResourcesPages(
pkg/destroy/aws/aws.go:	err2 := search.client.ListRolesPages(
pkg/destroy/aws/aws.go:	err2 := search.client.ListUsersPages(
pkg/destroy/aws/aws.go:	err2 := client.DescribeRouteTablesPages(
pkg/destroy/aws/aws.go:	err2 := client.DescribeLoadBalancersPages(
pkg/destroy/aws/aws.go:	err2 := client.DescribeTargetGroupsPages(
pkg/destroy/aws/aws.go:	err2 := client.DescribeLoadBalancersPages(
pkg/destroy/aws/aws.go:	err2 := client.ListRolePoliciesPages(
pkg/destroy/aws/aws.go:	err2 = client.ListInstanceProfilesForRolePages(
pkg/destroy/aws/aws.go:	err2 := client.ListUserPoliciesPages(
pkg/destroy/aws/aws.go:	err2 = client.ListAccessKeysPages(
pkg/destroy/aws/aws.go:	err = client.ListResourceRecordSetsPages(
pkg/destroy/aws/aws.go:	err2 := client.ListResourceRecordSetsPages(
```

It's unlikely that a single role will have pages of policies, but there could certainly be pages of private zones zone.  This non-paginated line would have lead to deeper zones being ignored, which may be the reason we'd sometimes leak private zones.  And that in turn would have lead to leaking records in the public zone.  And both of those leaks would require future destroy runs to grind through more Route 53 entries looking for things to delete, which was hurting our Route 53 API rate limits.

[1]: https://github.com/openshift/installer/pull/1039/files#r249244205

Reply:
Oops, I'd meant to comment on [this line][1] for the unpaginated-zones issue.

[1]: https://github.com/openshift/installer/blob/v0.10.0/pkg/destroy/aws/aws.go#L1388